docs/contributing.rst

```diff
@@ -71,12 +71,43 @@ or functions.
 Testing
 ^^^^^^^
 
-Twine is tested against Python versions 3.6, 3.7, and 3.8. To run these tests
-locally, you will need these versions of Python installed on your machine.
+We use `pytest`_ for writing and running tests.
+
+To run the tests in your virtual environment, run:
+
+.. code-block:: console
+
+  tox -e py
+
+To pass options to ``pytest``, e.g. the name of a test, run:
+
+.. code-block:: console
+
+  tox -e py -- tests/test_upload.py::test_exception_for_http_status
+
+You can also set the ``PYTEST_ADDOPTS`` environment variable to use the same
+options on every test run. For example, to skip integration tests and turn
+off the coverage report, run:
+
+.. code-block:: console
+
+  export PYTEST_ADDOPTS=-k "not integration" --cov-report=
+
+Twine is continuously tested against Python 3.6, 3.7, and 3.8 using `Travis`_.
+To run the tests against a specific version, e.g. Python 3.6, you will need it
+installed on your machine. Then, run:
+
+.. code-block:: console
+
+  tox -e py36
+
+To run the tests against all supported Python versions, check code style,
+and build the documentation, run:
```

Comment:
If I run `tox` inside a virtualenv, how can it test all supported python versions? 

I have python3.6, python3.7 and python3.8 installed on my MAC via pyenv, and when I ran `tox` within my virtualenv with python3.8, the tests for python3.6 and python3.7 failed with `ERROR: InterpreterNotFound:`. Am I missing something here?

Reply:
Nothing above suggests running tox from within a virtualenv. What made you think that was necessary?

Reply:
I was under the assumption that since we are using a virtualenv for development, we should be installing tox inside it and running. Is that an incorrect assumption?

Also on looking further,  I do see suggestion in the [twine contributing docs](https://twine.readthedocs.io/en/latest/contributing.html)  to run `tox` inside a virtualenv.

Quoting the docs:

In the Getting started section: `We use tox to run tests, check code style, and build the documentation. To install tox in your active your virtual environment, run:` (There is a typo here as well)

In the Building the documentation section: `Activate your virtual environment, then, in the root directory, run: tox -e docs`

Reply:
> To install tox in your active your virtual environment, run:

Yeah, we should rephrase that to:

```
If you prefer to have tox installed in a virtual environment, run:
```

Because while tox _can_ sometimes work just fine from within a virtualenv, doing so with that + pyenv (or asdf, or something else) will not work properly unless you're an expert with both virtualenv and the version manager of your choice. (This is in no small part due also to how those version managers choose to manage "active" or "available" versions - which is not useful for developing something against multiple versions of Python.)

Reply:
> This is in no small part due also to how those version managers choose to manage "active" or "available" versions - which is not useful for developing something against multiple versions of Python.

So what would you suggest using to develop against multiple versions of python. I currently have to switch python versions via `pyenv global <version>` and use that to create a new virtualenv, everytime I need to switch to another python version.

Reply:
@deveshks `tox` is the correct tool run the tests against multiple Python versions. I think you could avoid the `ERROR: InterpreterNotFound` by using `pyenv local` in the `twine` directory, e.g.:

```
$ pyenv local 3.6.10 3.7.3 3.8.2

$ cat .python-version 
3.6.10
3.7.3
3.8.2

$ pyenv which python3
/Users/brian/.pyenv/versions/3.6.10/bin/python3

$ pyenv which python3.6
/Users/brian/.pyenv/versions/3.6.10/bin/python3.6

$ pyenv which python3.7
/Users/brian/.pyenv/versions/3.7.3/bin/python3.7

$ pyenv which python3.8
/Users/brian/.pyenv/versions/3.8.2/bin/python3.8
```

Reply:
Thanks for that. I can run `tox` now in my virtualenv, and it tests all python versions (3.6, 3.7 and 3.8) as well.

Reply:
> Yeah, we should rephrase that to:
> 
> ```
> If you prefer to have tox installed in a virtual environment, run:
> ```

This remains a point of confusion for me. I see some projects recommending this (even including `tox` in their requirements), while others treat it as a generally-available command-line tool, independent of a particular virtual environment. I lean towards the latter, in which case the instructions would be something like `python3 -m pip install --user tox` and/or `pipx install tox` (which is what I use; see [pipx docs](https://pipxproject.github.io/pipx/)).